### PR TITLE
Multiple Relations Tests & Intra

### PR DESCRIPTION
**Pull request by MiaAltieri**
### Notes to the reader
1. This PR is large and updates a lot of files under `tests/integration/application` to be consistent with the new changes in [data-platform-libs](https://github.com/canonical/data-platform-libs/).
2. The integration tests are simply the same tests from the [data-platform-libs](https://github.com/canonical/data-platform-libs/blob/main/tests/integration/test_charm.py), just fitted to MongoDB

### Problem

Application charms using the requires charm library should be able to connect to multiple database clusters through the same relation.

### Solution

This PR tests relations with MongoDB for:
1. multiple apps connecting to the same database have unique credentials 
2. a single application can connect to multiple databases, such that these connections are unique (ie unique credentials)
3. an application can re-use the same interface to request multiple databases

### Context

The app-data visible via `show-unit` is the remote app data, ideally we could see all relevant app data (ie the local app data as well). This idea is maintained in the current [data-platform-lib](https://github.com/canonical/data-platform-libs/blob/main/lib/charms/data_platform_libs/v0/database_provides.py) implementation. To keep things consistent across the DPE charms, we utilise the `data-platform-libs` `diff` function so that the MongoDB charm updates its app data to be consistent with its related app data.

### Release Notes

1. Added tests to verify application can be related to multiple databases 
2. Updated `mongodb_provider.py` to use a `_diff` function to update its relation app data, as to be consistent with the patterns from the [data-platform-lib](https://github.com/canonical/data-platform-libs/blob/main/lib/charms/data_platform_libs/v0/database_provides.py) 

### Testing

Integration tests added
Unit tests will be added in a future PR

### CODE REVIEW

File: tests/integration/relation_tests/helpers.py [hunk 1]
```diff
@@ -1,35 +1,64 @@
...
     unit_name = f"{application_name}/0"
     raw_data = (await ops_test.juju("show-unit", unit_name))[1]
+
     if not raw_data:
         raise ValueError(f"no unit info could be grabbed for {unit_name}")
     data = yaml.safe_load(raw_data)
-    return data[unit_name]["relation-info"][0]["application-data"].get(key)
+
+    # Filter the data based on the relation name.
+    relation_data = [v for v in data[unit_name]["relation-info"] if v["endpoint"] == relation_name]
```
Comment (WRFitch) [inline tests/integration/relation_tests/helpers.py:45]: I'm a little confused about the meaning of `v` - does it just mean `variable`?

File: tests/integration/relation_tests/test_charm_relations.py [hunk 2]
```diff
@@ -15,7 +15,12 @@
 DATABASE_METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
 PORT = 27017
 DATABASE_APP_NAME = DATABASE_METADATA["name"]
-APP_NAMES = [APPLICATION_APP_NAME, DATABASE_APP_NAME]
+FIRST_DATABASE_RELATION_NAME = "first-database"
+SECOND_DATABASE_RELATION_NAME = "second-database"
+MULTIPLE_DATABASE_CLUSTERS_RELATION_NAME = "multiple-database-clusters"
+ALIASED_MULTIPLE_DATABASE_CLUSTERS_RELATION_NAME = "aliased-multiple-database-clusters"
+ANOTHER_DATABASE_APP_NAME = "another-database"
+APP_NAMES = [APPLICATION_APP_NAME, DATABASE_APP_NAME, ANOTHER_DATABASE_APP_NAME]
```
Comment (WRFitch) [inline tests/integration/relation_tests/test_charm_relations.py:23]: If these constants are going to be reused, we could open a pr to data-platform-libs to make them constants in test_charm.py, and you could import them from there.

File: tests/integration/relation_tests/test_charm_relations.py [hunk 3]
```diff
@@ -91,7 +109,132 @@ async def test_user_with_extra_roles(ops_test: OpsTest):
...
+    # Deploy another application.
+    await ops_test.model.deploy(
+        application_charm,
+        application_name=another_application_app_name,
+    )
+    await ops_test.model.wait_for_idle(apps=all_app_names, status="active")
+
+    # Relate the new application with the database
+    # and wait for them exchanging some connection data.
+    await ops_test.model.add_relation(
```
Comment (WRFitch) [inline tests/integration/relation_tests/test_charm_relations.py:134]: To ensure your tests are independent of each other, you could remove your relations after each test; this way, they aren't a confounding variable in the following tests, which would be hard to debug.

**Review by WRFitch — APPROVED** [review]
LGTM

File: tests/integration/relation_tests/helpers.py (hunk 1 above)
Comment (jnsgruk) [reply tests/integration/relation_tests/helpers.py:45]: https://www.w3schools.com/python/python_lists_comprehension.asp :)

File: tests/integration/relation_tests/test_charm_relations.py (hunk 3 above)
Comment (MiaAltieri) [reply tests/integration/relation_tests/test_charm_relations.py:134]: PMing you!!

File: lib/charms/mongodb_libs/v0/mongodb_provider.py [hunk 4]
```diff
@@ -67,13 +78,13 @@ def _on_relation_event(self, event):
...
         try:
-            self.oversee_users(departed_relation_id)
+            self.oversee_users(departed_relation_id, event)
         except PyMongoError as e:
             logger.error("Deferring _on_relation_event since: error=%r", e)
             event.defer()
             return
 
-    def oversee_users(self, departed_relation_id: Optional[int]):
+    def oversee_users(self, departed_relation_id: Optional[int], event):
```
Comment (paulomach) [inline lib/charms/mongodb_libs/v0/mongodb_provider.py:87]: Nit: add `event` to docstring

File: tests/integration/relation_tests/application-charm/metadata.yaml [hunk 5]
```diff
@@ -8,5 +8,12 @@ summary: |
...
-  database:
+  first-database:
     interface: mongodb-client
+  second-database:
+    interface: mongodb-client
+  multiple-database-clusters:
+    interface: mongodb-client
+  aliased-multiple-database-clusters:
+    interface: mongodb-client
+    limit: 2
```
Comment (paulomach) [inline tests/integration/relation_tests/application-charm/metadata.yaml:19]: nit: new line

**Review by paulomach — APPROVED** [review]
LGTM

File: lib/charms/mongodb_libs/v0/mongodb_provider.py [hunk 6]
```diff
@@ -110,6 +121,8 @@ def oversee_users(self, departed_relation_id: Optional[int]):
                 config = self._get_config(username)
                 logger.info("Update relation user: %s on %s", config.username, config.database)
                 mongo.update_user(config)
+                logger.info("Updating relation data according to diff")
+                self._diff(event)
```
Comment (marceloneppel) [inline lib/charms/mongodb_libs/v0/mongodb_provider.py:125]: Should the result of this function be handled in some way on MongoDB side (does it use the diff)?

**Review by marceloneppel — APPROVED** [review]
LGTM! I have just one question about the use of the `diff` inside the provider library.

File: lib/charms/mongodb_libs/v0/mongodb_provider.py (hunk 6 above)
Comment (MiaAltieri) [reply lib/charms/mongodb_libs/v0/mongodb_provider.py:125]: I don't think it directly uses the diff, but I think the diff function issues a relation changed event when it does:
`event.relation.data[self.charm.model.app].update({"data": json.dumps(new_data)})`, which triggers the mongodb code to handle the diff, idk if that makes sense or not.

**Pull request merged**